src/HamiltonianBuilder/FCI.cpp

```diff
@@ -0,0 +1,562 @@
+                                    }  // create on s (beta)
+                                }  // loop over s
+
+                                spin_string_beta.annihilate(r);  // undo the previous creation on r
+                            }  // create on r (beta)
+                        }  // loop over r
+
+                        spin_string_beta.create(q);  // undo the previous annihilation on q
+                    }  // annihilate on q (beta)
+                }  // loop over q
+
+                spin_string_beta.create(p);  // undo the previous annihilation on p
+            } // annihilate on p (beta)
+        }  // loop over p
+
+        if (I_beta < dim_beta - 1) {  // prevent last permutation to occur
+            fock_space_beta.setNext(spin_string_beta);
+        }
+    }  // loop over beta addresses (I_beta)
+
+
+    // 3. ALPHA-BETA
+    for (size_t I_alpha = 0; I_alpha < dim_alpha; I_alpha++) {  // loop over alpha addresses
+        for (size_t I_beta = 0; I_beta < dim_beta; I_beta++) {  // loop over beta addresses
+
+            for (const auto& alpha : this->alpha_one_electron_couplings[I_alpha]) {  // traverse all OneElectronCouplings for I_alpha
+                for (const auto& beta : this->beta_one_electron_couplings[I_beta]) {  // traverse all OneElectronCouplings for I_beta
+
+                    int sign = alpha.sign * beta.sign;
+                    double value = sign * hamiltonian_parameters.get_g().get(alpha.p, alpha.q, beta.p, beta.q);
+                    result_matrix( I_alpha * dim_beta + I_beta, alpha.address * dim_beta + beta.address) += value;  // alpha is the major index
+                }  // beta OneElectronCouplings
+            }  // alpha OneElectronCouplings
+
+        }  // loop over beta addresses (I_beta)
+    }  // loop over alpha addresses (I_alpha)
+    return result_matrix;
+}
+
+
+/**
+ *  @return the action of the Hamiltonian (@param hamiltonian_parameters and @param diagonal) on the coefficient vector @param x
+ */
+Eigen::VectorXd FCI::matrixVectorProduct(const HamiltonianParameters& hamiltonian_parameters, const Eigen::VectorXd& x, const Eigen::VectorXd& diagonal) {
+    auto K = hamiltonian_parameters.get_h().get_dim();
+    if (K != this->fock_space.get_K()) {
+        throw std::invalid_argument("Basis functions of the Fock space and hamiltonian_parameters are incompatible.");
+    }
+
+    FockSpace fock_space_alpha = fock_space.get_fock_space_alpha();
+    FockSpace fock_space_beta = fock_space.get_fock_space_beta();
+
+    auto dim_alpha = fock_space_alpha.get_dimension();
+    auto dim_beta = fock_space_beta.get_dimension();
+    auto dim = fock_space.get_dimension();
+
+    // TODO: use diagonal
+    Eigen::VectorXd matvec =  Eigen::VectorXd::Zero(dim);
+
+    // Calculate the effective one-electron integrals
+    // TODO: move this to libwint
+    Eigen::MatrixXd k_SO = hamiltonian_parameters.get_h().get_matrix_representation();
```

Comment:
I made an issue #52 to put this function into a function of `HamiltonianParameters`.